### ISSUE TITLE
Edit notifications don't work when running inside of Docker container on Windows

### ISSUE THREAD

**Issue by matt-github-acct** [fsnotify/fsnotify#292]
I'm running Windows 10 Enterprise, 10.0.17134

My goal is to launch a docker container, mount a volume inside the container, edit files in the volume from my host operating system and have the docker container recognize the change events. Pretty much this: https://www.zachjohnsondev.com/posts/go-docker-hot-reload-example/

*Steps to reproduce*
1. Create `main.go` file and add the following:
```
package main

import (
	"log"

	"github.com/fsnotify/fsnotify"
)

func main() {
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		log.Fatal(err)
	}
	defer watcher.Close()

	done := make(chan bool)
	go func() {
		for {
			select {
			case event, ok := <-watcher.Events:
				if !ok {
					return
				}
				log.Println("event:", event)
				if event.Op&fsnotify.Write == fsnotify.Write {
					log.Println("modified file:", event.Name)
				}
			case err, ok := <-watcher.Errors:
				if !ok {
					return
				}
				log.Println("error:", err)
			}
		}
	}()

	err = watcher.Add(".")
	if err != nil {
		log.Fatal(err)
	}
	<-done
}

```
1. Create a `docker-compose` file and add the following:
```
version: "3.6"

services:
  bar:
    container_name: bar
    build: .
    ports:
      - "8080:8080"
    volumes:
      - ./:/go/src/bar
```
1. Create a `Dockerfile` file and add the following:
```
FROM golang:latest

WORKDIR /go/src/bar

COPY . .

RUN ["go", "get", "github.com/fsnotify/fsnotify"]

ENTRYPOINT go run main.go
```
1. Run `docker-compose build` in the directory where you've created the files.
1. Run `docker-compose up` in the directory where you've created the files.
1. From your host operating system, add a comment to `main.go` and notice that no edit events are reported.
1. Run `docker exec -it bar bash` to connect to the docker container. Edit `main.go` from the container shell to confirm that edit events are being reported when the edit is performed within the container.

**Comment by oxycoder** [comment]
Same problem, any one have solutions?

**Comment by nathany** [comment]
I'm not sure how Docker for Windows is mounting the shared folder, but generally file shares such as NFS don't support notifications.

A polling fallback (#9) is the only solution I know of when the underlying system doesn't provide events.

For now we should document the limitation, but first we should confirm on other operating systems. It may work on Linux, depending on how the Docker container is mounted there.

**Comment by oxycoder** [comment]
Currently it work perfectly fine with folder mount to docker on MacOS.

**Comment by julienkosinski** [comment]
I can confirm it works well under on Docker with Linux host (Manjaro distro)

**Comment by jnach** [comment]
> I can confirm it works well under on Docker with Linux host (Manjaro distro).

Which storage driver do you happen to be using? With @nathany's comment in mind, this could be filesystem dependent. In both 'working' cases, this sounds like a bind mount, not a real storage driver under the PVC subsystem

**Comment by jnach** [comment]
[This](https://william-yeh.net/post/2019/06/inotify-in-containers/) seems to do a good job of explaining the phenomenon here..

**Comment by jnach** [comment]
> Same problem, any solutions?

Sorta, see above. TL&DR This is a file-system compatibility issue. There are a few scenarios where it works when using the _same underlying host filesystem_, but pretty much never if there is any translation.

**Comment by reyronald** [comment]
I'm also interested in a workaround or solution to this, if there's any. Would be ok with polling if that does the job. Could be a CLI option

**Comment by Hanggi** [comment]
Me too,

Is there any certain plan to cover this issue? 

There are several packages using fsnotify (eg: air, compileDaemon ...), 
I have tried them all, and these packages have the same issue, certainly.

**Comment by arp242** [comment]
Closed as duplicate of #9; if the underlying OS mechanism doesn't work then there's not much we can do about it other than implement a polling fallback.
